File: 2018/day05.py
import fileinput
import string


def collapse(polymer):
    units = []

    for c in polymer:
        units.append(c)
        while len(units) > 1 and abs(ord(units[-1]) - ord(units[-2])) == 32:
            units.pop()
            units.pop()

    return "".join(units)


def p1(polymer):
    result = collapse(polymer)

    return len(result)


def p2(polymer):
    collapsed = collapse(polymer)

    candidates = [
        collapse([x for x in collapsed if x not in {c, chr(ord(c) + 32)}])
        for c in string.ascii_uppercase
    ]
    return min(len(x) for x in candidates)


if __name__ == "__main__":
    polymer = next(iter(fileinput.input())).strip()
    print("Part 1: %d" % p1(polymer))
    print("Part 2: %d" % p2(polymer))

File: 2018/cpp/day05.cc
#include <iostream>
#include <string>
#include <vector>

#include "utils.h"

std::vector<char> collapse(const std::vector<char> &polymer) {
  std::vector<char> result;
  for (auto c : polymer) {
    while (result.size() and std::abs(result.back() - c) == 32) {
      result.pop_back();
      c = result.back();
      result.pop_back();
    }
    result.push_back(c);
  }

  return result;
}

int p1(const std::string &polymer) {
  auto collapsed = collapse({polymer.begin(), polymer.end()});
  return collapsed.size();
}

int p2(const std::string &polymer) {
  auto collapsed = collapse({polymer.begin(), polymer.end()});
  std::vector<int> lengths;

  for (char c = 'A'; c <= 'Z'; c++) {
    std::vector<char> candidate = collapsed;
    for (auto it = candidate.begin(); it != candidate.end();) {
      if (*it == c || *it == std::tolower(c)) {
        it = candidate.erase(it);
      } else {
        ++it;
      }
    }
    lengths.push_back(collapse(candidate).size());
  }

  return *std::min_element(lengths.begin(), lengths.end());
}

int main() {
  std::vector<std::string> lines = aoc::utils::getlines();
  std::sort(lines.begin(), lines.end());

  std::cout << "Part 1: " << p1(lines[0]) << std::endl;
  std::cout << "Part 2: " << p2(lines[0]) << std::endl;
}
